### ISSUE TITLE
No history, bookmarks etc. preserved in lynx browser

### ISSUE THREAD

**Issue by Rosika2** [netblue30/firejail#3570]
**Environment**
 - Linux distribution and version: Lubuntu 18.04.4 LTS, 64 bit
 - Firejail version 0.9.62
 - What other programs interact with the affected program for the functionality? lynx
 - Are these listed in the profile? lynx.profile
----------------------------------------------------------------------------------------------------------------------------------------------------

Hi altogether,

I´ve got a question concerning the use of **lynx** (terminal-browser) and firejail with the _private option_.

Wanting to use firejail **with a defined work directory** I use the command

```
firejail --private=/home/rosika/Dokumente/kgw/kgw2 lynx

Reading profile /etc/firejail/lynx.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-passwdmgr.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/disable-xdg.inc
Warning: networking feature is disabled in Firejail configuration file
Parent pid 6365, child pid 6366
Warning: cleaning all supplementary groups
Warning: cleaning all supplementary groups
Warning: skipping private-cache: cannot find /home/rosika/.cache
Blacklist violations are logged to syslog
Child process initialized in 5911173136384.00 ms

Parent is shutting down, bye...

```

It works well but when shutting down lynx no search history, bookmarks or cookies are preserved.
I know that´s the case with "--private" alone but I added a work-directory. So theoretically those items should be saved I guess.

Besides in the kgw2-folder only two items are created: **.Xauthority** and **.bashrc**.

As a second step I used the command 

```
firejail --profile=/etc/firejail/default.profile --private=/home/rosika/Dokumente/kgw/kgw2 lynx  

Reading profile /etc/firejail/default.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-passwdmgr.inc
Reading profile /etc/firejail/disable-programs.inc
Warning: networking feature is disabled in Firejail configuration file
Parent pid 6416, child pid 6417
Warning: cleaning all supplementary groups
Warning: cleaning all supplementary groups
Warning: cleaning all supplementary groups
Child process initialized in 59.86 ms

Parent is shutting down, bye...


```

to see if there´s any difference.

In my private folder there´s one additional entry: besides .Xauthority and .bashrc there´s a folder called **.config**.
In it there are the folders **pulse** and **fish**. Pulse is empty and fish just contains the file "fishd.rosika-Lenovo-H520e".

So no lynx history, bookmarks or config-file either.

What am I doing wrong here? Could you help me in some way?

Thanks a lot in advance.

Greetings.
Rosika

**Comment by kortewegdevries** [comment]
>In my private folder there´s one additional entry: besides .Xauthority and .bashrc there´s a folder called .config.
..........
So no lynx history, bookmarks or config-file either.

What do you want to do? Retain changes when it is run inside the private directory after closing or use your current configuration inside private-directory?

**Comment by Rosika2** [comment]
@kortewegdevries:

Hi and thank you for your reply.

> What do you want to do? Retain changes when it is run inside the private directory after closing

Yes. That´s basically what I want to achieve.

When running my main browser "chromium-browser" I use this command:
`firejail --private=/home/rosika/Dokumente/work chromium-browser `
and all my bookmarks, history etc. are preserved.

I assumed this shoud work with lynx as well.

Greetings.
Rosika

**Comment by rusty-snake** [comment]
It should work as well with lynx. I have absolutely no idea what causes this. @Rosika2 can you run `firejail --noprofile --private=/home/rosika/CLEAN/PRIVATE/DIR lynx` to see where they are stored (in the hope that  this stores it).

**Comment by kortewegdevries** [comment]
> I assumed this should work with lynx as well.

This is working for me. I'm using`firejail --private=/home/korte/TempDir /usr/bin/lynx wikipedia.com` Lynx saves all bookmarks in `lynx_bookmarks.html` and it stays there.

**Comment by Rosika2** [comment]
@rusty-snake and @kortewegdevries:

Hi again and thanks a lot for your answers,

alas none of your suggestions work for me.

I tried both of you solution commands and in both cases I get exactly the same results as before:
Those files and folders are created:

- .Xauthority and .bashrc (files)
- .config (folder)
- within .config: pulse and fish (folders)
- pulse: empty
- fish: fishd.rosika-Lenovo-H520e

After closing lynx all bookmarks are lost.
Very strange.

Greetings.
Rosika

**Comment by kortewegdevries** [comment]
Are you using a different shell other than bash? Is the bookmark file being created?

**Comment by Rosika2** [comment]
Hi again,

yes, I use fish as my default shell.
But I tried the commands in bash as well. The results are the same.

No idea if it helps at all but I also noticed something else which seems a bt off:

Using the command
`firejail --noprofile --private=/home/rosika/CLEAN/PRIVATE/DIR lynx -dump "https://itsfoss.community/t/most-recent-grub2-updates-rendering-some-systems-unbootable/5146/3" > ziel.txt`
the resulting text-file **DOESN´T** appear in _/home/rosika/CLEAN/PRIVATE/DIR_ .

However  I found it in _/home/rosika_ !!!!!!!!

**Comment by Rosika2** [comment]
New info:

I just ran lynx **outside** firejail, just as a normal command.

It turns out that lynx doesn´t save bookmarks that way either. So it seems it´s a lynx issue.
Never would have thought that.

Curious though that @kortewegdevries says:

"This is working for me. I'm usingfirejail --private=/home/korte/TempDir /usr/bin/lynx wikipedia.com Lynx saves all bookmarks in lynx_bookmarks.html,.lynxrc and it stays there."

@kortewegdevries: 
How do you do it? Do you use any special settings?

**Comment by Rosika2** [comment]
Hi,

my previous comment:

```
Using the command
firejail --noprofile --private=/home/rosika/CLEAN/PRIVATE/DIR lynx -dump "https://itsfoss.community/t/most-recent-grub2-updates-rendering-some-systems-unbootable/5146/3" > ziel.txt
the resulting text-file DOESN´T appear in /home/rosika/CLEAN/PRIVATE/DIR .
However I found it in /home/rosika !!!!!!!!
```
was a bit unqualified I think and seems to be rooted in a misconception on my part. Sorry.
It´s clear that the text-file should be saved in my home directory as I started the command from there.

But I noticed the following:

Usng my locate-command  I couldn´t find **lynx_bookmarks.html** or **.lynxrc** files anywhere on my system. They simply don´t seem to exist.

**Comment by rusty-snake** [comment]
Is the history disable in `/etc/lynxrc`?

**Comment by kortewegdevries** [comment]
> Usng my locate-command I couldn´t find lynx_bookmarks.html or .lynxrc files anywhere on my system. They simply don´t seem to exist.

I forgot, open a link press "A">"L" save?

Is the history disabled in /etc/lynxrc?

Supposed to be in `/home/<>` or you mean `lynx.cfg`?

**Comment by Rosika2** [comment]
@rusty-snake:
Sorry, I haven´t got a lynxrc file in /etc.

I have the following:

```
locate lynxrc

/home/rosika/.local/share/Trash/files/.2.lynxrc
/home/rosika/.local/share/Trash/files/.lynxrc
/home/rosika/.local/share/Trash/info/.2.lynxrc.trashinfo
/home/rosika/.local/share/Trash/info/.lynxrc.trashinfo
```

**Comment by kortewegdevries** [comment]
My file if you need it: lynx_bookmarks.html https://termbin.com/9k7s

> I forgot, open a link press "A">"L" save?

Did this work?

**Comment by Rosika2** [comment]
@kortewegdevries:
Sorry, I have to ask further.
What exactly do you mean, what should I do?

**Comment by kortewegdevries** [comment]
To save a bookmark open a link on lynx press "A" the press "d" or "L" or "c" then enter? That's how I'm saving it and "enter" creates that file?

**Comment by Rosika2** [comment]
@kortewegdevries:
Thanks.
I´m not so sure about the bookmarks but I think the reason why the **history** isn´t saved is that it´s transferred to **/tmp**.
Lynx says:
```
Linkadressenliste (Lynx Version 2.8.9dev.16), help

   Verweise in file://localhost/tmp/lynxXXXX6gZvqf/L5-2723TMP.html
```
And when I take a look in **/tmp** I find it: 
```
rosika@rosika-Lenovo-H520e /t/lynxXXXX6gZvqf> tree
.
├── L5-4811TMP.html
└── L5-4948TMP.html
```
Here is my history. I looked it up and every site I visited is stored in **L5-4811TMP.html**.
But as it is in  **/tmp** it´s not available next time I use lynx of course.

**Comment by kortewegdevries** [comment]
No idea, I don't use lynx sory. But man firejail says 

> --private=directory
> Use directory as user home.  --private and --private=directory cannot be used together.

So `/tmp' stays out of your temporary directory. Maybe try this recent:

**Comment by Rosika2** [comment]
Hi,
> So bookmarks get saved?

Didn´t manage to save any bookmarks.
Perhaps I´m doing it wrong, despite your instructions. Yet I´ll keep trying and report back.

Thanks.

**Comment by Rosika2** [comment]
O.K. I tried again and lynx says something like "setting bookmarks [...etc] not possible".

I really think all the problems might be rooted within lynx itself as running lynx outside the sandbox doesn´t preserve history either.

**Comment by kortewegdevries** [comment]
Can you like set it without firejail or with --noprofile? Press "A" then "L" then Enter.

**Comment by Rosika2** [comment]
Well I´ve been using --noprofile all along:
`firejail --noprofile --private=/home/rosika/CLEAN/PRIVATE/DIR lynx
`
And lynx explictely says that setting bookmarks isn´t possible.

**Comment by kortewegdevries** [comment]
No idea, you can have my configs:
```
korte:~$ cat /etc/lynx/lynx.cfg|nc termbin.com 9999
https://termbin.com/ff28
korte:~$ cat /etc/lynx/lynx.cfg|nc termbin.^Cm 9999
korte:~$ ls /etc/lynx/
lynx.cfg  lynx.lss
korte:~$ cat /etc/lynx/lynx.lss|nc termbin.com 9999
https://termbin.com/50xx
```
Local configs are likely stored in `.lynxrc` I think, but I don't have it.

**Comment by Rosika2** [comment]
@kortewegdevries:
Thank you so much. I´ll look into it.

Strange:
I could create bookmarks now!
And a **lynx_bookmarks.html** file is created in the private directory. That´s working now.

But: when shutting down lynx and then re-starting it lynx doesn´t find lynx_bookmarks.html despite the fact it´s there.

I suppose it loooks somewhere else for that file.

**Comment by kortewegdevries** [comment]
This issue is too weird! I can read the bookmarks even when restarting under firejail+private: https://imgur.com/BwpEckT

About history I don't know where they are saved but if it's in `/tmp` like you say maybe you can set up a temporary temp directory. About cookies, same as above. About local settings maybe try and copy it to your temporary home so that lynx can read it,

**Comment by Rosika2** [comment]
@kortewegdevries:

> This issue is too weird!

Yes, you´re right. Never mind. Thanks a lot anyway for your help.

Oh, believe it or not, I´ve just managed to save bookmarks.
Out of sheer desperation I tried the following:

Consdering it a bit curious that lynx puts its bookmarks file under ` /home/rosika/CLEAN/PRIVATE/DIR/` instead of `.config` I transferred the bookmark file to` /home/rosika/CLEAN/PRIVATE/DIR/.config` and changed the respective bookmarks path within lynx (pressing "o") to that path.

And now it works.

BTW:my command is:` env LANG=en_GB:en firejail --noprofile --private=/home/rosika/CLEAN/PRIVATE/DIR /usr/bin/lynx`
The env LANG part is just for the sake of getting the English  language variant.

So bookmarks are working now. Yet ´ve still to figure out how I can get history to be saved as well.

Greetings.
Rosika

**Comment by Rosika2** [comment]
Hi again,

I **almost** solved the problem of how to save history.

I created a file `/home/rosika/CLEAN/PRIVATE/DIR/.config/.lynx_session/` and within lynx:
typing "o" and then under "Special Files and Screens" entered the respective path in "Session file". 
Plus: turning "Auto Session" on.

That would work! The history is stored in **lynx_session**.

BUT:   Auto Session (!)  and   Session file (!)   have exclamation marks which means: "(options marked with (!) will not be saved)".
I.e.: when starting lynx anew the lynx_session file is still there but it is NOT accessed.
Any ideas how to make the changes in settings persist?

Greetings.
Rosika

**Comment by kortewegdevries** [comment]
Lynx has this, maybe try:

```
 -cookie_file=FILENAME
                    specifies a file to use to read cookies
  -cookie_save_file=FILENAME
                    specifies a file to use to store cookies
  -cookies    toggles handling of Set-Cookie headers (on)
```
I'll check on the history.

**Comment by Rosika2** [comment]
@kortewegdevries:

Hello again and thank you so much for your help. You´re a genius.

SUCCESS!!

It finally worked.

> Lynx can save and restore useful information about your browsing history. Use this setting to enable or disable the feature.

This put me on the right track.
I looked up the "session"-entries in the man-pages and found this:

"−session=FILENAME
resumes from specified file on startup and saves session to that file on exit."

So I put the **.lynx_session** file into `/home/rosika/CLEAN/PRIVATE/DIR/` and started lynx with the command
`env LANG=en_GB:en firejail --name=mysandbox --private=/home/rosika/CLEAN/PRIVATE/DIR lynx -session=.lynx_session
`
That way the **.lynx_session** file is read every time lynx starts and updated every time lynx is shut down.
Wonderful!

Now history is saved and bookmarks can be created as well.

Plus: everything works now as it´s supposed to with firejail (using the --private option with defined work-directory).

I´m so glad and thankful for your persistent help.
Thanks for not giving up on me.  :)

Many greetings.
Rosika

P.S.:
> (The bug tag should be removed.)

I looked everywhere but couldn´t find a way to do this.
